Erweitern der Siegbedingungen

BroderPeters opened the pull request:
Infos stehen gut zusammengefasst in den Commits.

Commit:
[#163] expand winning-conditions

add "game.player.AntCount" to avoid multiple calculations
add "checkMoneyAndAnts()" to "checkWinningConditions()"
remove "this." from "this.stop()" in "GamePanel.cs"
add "checkMoneyAndAnts()" to check wether money and ants from a player are near zero
add "PLAYER_WON_OUT_OF_MONEY_AND_ANTS" to message a winning statement
add "LOWEST_COST_VALUE" and "getLowestCostValue()" to calculate the lowest costvalue in "Helper.CostCalculator.cs"
add "lbl_attackPower" and "nmc_warriorAttackPower" in "CostCalculator.cs" to calculate costs in windows form

Commit:
moves and formats costvalues in "CostCalculator.cs"

Commit:
add summary for "Player.AntCount"

jmattheis commented on AntWars/Player.cs:
```diff
@@ -68,10 +70,13 @@ class Player {
         public void incrementAnts(Ant ant) {
             if (ant.isCarry()) {
```
Du kannst vor das if einfach `AntCount++;` machen dann musst du es nicht 3x haben das gleiche für `DeathCount`

BroderPeters replied:
Dann würde es aber auch hochgezählt werden, wenn der else-Zweig mit der Exception auftritt.
Macht zwar keinen Unterschied, aber ich fand es so sauberer.

jmattheis replied:
wenn ne exception da auftritt, bricht das programm ab, so hast du halt unnötigerweile 3x die gleiche zeile die auch einfach an den anfang kann, da die methode ja increaseAnts heißt, der else zweig ist hauptsächlich dazu da, das man wenn man einen neuen ameisen typ hinzufügt, es da nicht vergisst hinzuzufügen.

jmattheis commented on AntWars/Player.cs:
```diff
@@ -44,6 +44,8 @@ class Player {
         /// </summary>
         public int CarryCount { get; private set; }
 
+        public int AntCount { get; private set; }
```
dokú

jmattheis commented on AntWars/Helper/Messages.cs:
```diff
@@ -96,5 +96,10 @@ static class Messages {
         public static readonly string VIEWRANGE = "Sichtweite";
+
+        /// <summary>
+        /// {0} hat gewonnen! {1} hat weder genügend Ameisen noch genügend Geld.
```
"hat weder genügend Ameisen noch genügend Geld." klingt nicht richtig, in welchem kontext nicht "genügend". sollte sowas sein wie hat keine ameisen und kein geld für neue.

jmattheis commented on AntWars/Helper/CostCalculator.cs:
```diff
@@ -111,5 +113,18 @@ class CostCalculator {
+                return lowCarry;
+            else if (lowScout <= lowCarry && lowScout <= lowWarrior)
+                return lowScout;
+            else // if (lowWarrior < lowCarry && lowWarrior < lowScout), Auskommentiert da alle Codepfade einen Wert zurückgeben müssen.
```
der kommentar kann weg

jmattheis commented on AntWars/Helper/CostCalculator.cs:
```diff
@@ -111,5 +113,18 @@ class CostCalculator {
             return Math.Round((numbers.Sum() / 2), 2);
         }
+
+        public static double getLowestCostValue()
```
kannst du private machen weils ja blos in der eigenen klasse benutzt wird

jmattheis commented on AntWars/Helper/CostCalculator.cs:
```diff
@@ -7,6 +7,8 @@ namespace AntWars.Helper {
 
     class CostCalculator {
 
+        public static readonly double LOWEST_COST_VALUE = getLowestCostValue();
```
doku

jmattheis commented:
nochmal zu den commit `add "lbl_attackPower" and "nmc_warriorAttackPower" in "CostCalculator.cs" to calculate costs in windows form` du sollst nicht genau beschreiben was du gemacht hast, wenn man das wissen möchte kann man sich auch die code änderungen anschauen, es könnte zb sowas sein wie `add missing attackPower numericfield to the costcalculator`

BroderPeters commented:
Ist die Gewöhnung von Arbeit.

Commit:
implements improvments from pull request

increments and decrease "AntCount" in front of if-statements
change message "PLAYER_WON_OUT_OF_MONEY_AND_ANTS"
remove command
change "getLowestCostValue()" from public to private
add documentation

BroderPeters commented:
@jmattheis wurde behoben

Commit:
change summary for LOWEST_COST_VALUE

Commit:
add brackets for conventions

Commit:
add brackets for convention

jmattheis commented on AntWars/GamePanel.cs:
```diff
@@ -193,17 +199,33 @@ partial class GamePanel : Form {
         }
 
+        private bool checkMoneyAndAnts() {
+            if (game.Player1.Money < Helper.CostCalculator.LOWEST_COST_VALUE && game.Player1.AntCount == 0)
```
in Methode auslagern die anderen auch

Commit:
split functions for checking winning conditions

Commit:
Merge online changes with local and split functions for winning conditions

Commit:
remove forgotten merge conflict

BroderPeters commented:
@jmattheis @nicolasries Mergebar?

jmattheis commented on AntWars/GamePanel.cs:
```diff
@@ -183,47 +183,70 @@ partial class GamePanel : Form {
+            return false;
+        }
+
+        private bool checkMoneyAndAntsByPlayer(Player player) {
```
du könntest hier `private bool checkMoneyAndAntsByPlayer(Player player, Player other) {` machen dann geweils so aufrufen 

```
checkMoneyAndAntsByPlayer(game.Player2, game.Player1);
checkMoneyAndAntsByPlayer(game.Player1, game.Player2);
```

jmattheis replied:
dann brauchst du getOtherPlayer nicht

jmattheis commented on AntWars/GamePanel.cs:
```diff
@@ -183,47 +183,70 @@ partial class GamePanel : Form {
+        }
+
+        private bool checkMoneyAndAnts() {
+            if (checkMoneyAndAntsByPlayer(game.Player1)) {
```
return checkMoneyAndAntsByPlayer(game.Player1) || checkMoneyAndAntsByPlayer(game.Player2);

jmattheis commented on AntWars/GamePanel.cs:
```diff
@@ -183,47 +183,70 @@ partial class GamePanel : Form {
         }
 
         private bool checkMaxPoints() {
-            if (game.Player1.Points >= game.Conf.Points) {
```
kann auch ein einzeilersein wie checkMoenyAndAnts

jmattheis commented on AntWars/GamePanel.cs:
```diff
@@ -183,47 +183,70 @@ partial class GamePanel : Form {
-                stop();
-                MessageBox.Show(String.Format(Messages.PLAYER_WON_TIME_OUT,
-                                game.Player1.AI.Playername, game.Player1.Points), Messages.PLAYER_WON_CAPTION, MessageBoxButtons.OK, MessageBoxIcon.Information);
+            if (checkPlayerPointsByPlayer(game.Player1)) {
```
kann auch ein einzeilersein wie checkMoenyAndAnts

Commit:
perform improvments

Commit:
perform improvments

Commit:
solves bracket error

jmattheis commented on AntWars/Helper/CostCalculator.cs:
```diff
@@ -111,5 +116,19 @@ class CostCalculator {
+            double lowCarry = calculateCostCarry(1, 1, 1, 1, "");
+            double lowScout = calculateCostScout(1, 1, 1, 1, "");
+            double lowWarrior = calculateCostWarrior(1, 1, 1, 1, 1, "");
+            if (lowCarry <= lowScout && lowCarry <= lowWarrior) {
```
Sorry ist mir erst jetzt aufgefallen, es mit `Math.Min(x, y)` welcher dir den nidrigsten wert gibt,

```
double tmpMin = Math.Min(lowCarry, lowScout);
return Math.Min(tmpMin, lowWarrior);
```

jmattheis commented:
sonst noch mit master mergen, und dann hast du noch formatting dinge die nicht nach unserem standart sind, ich würde dann bei mir einmal den formatter durchjagen wenn du das obere noch gefixt hast.

Commit:
use Math#Min for getting the lowest value

Commit:
Merge branch 'master' into feature/winning_condition

jmattheis commented:
muss mal nochmal den master branch pull'en und dann nochmal mergen

Commit:
Merge branch 'master' into feature/winning_condition

Commit:
remove mergeconflict from Ai.cs

jmattheis commented:
btw beim mergen, sollen die conflicte gleich mit in dem merge commit sein, dann wird in die commit nachricht automatisch die datein die konflilte haben mit reingeschrieben

BroderPeters commented:
Habe den Mergeconflikt nicht gesehen, da er in der AI.cs war

jmattheis commented:
hey noch nicht approved gesagt ^^
